dgl_sparse/src/reduction.cc

```diff
@@ -0,0 +1,101 @@
+/**
+ *  Copyright (c) 2022 by Contributors
+ * @file reduction.cc
+ * @brief DGL C++ sparse matrix reduction operator implementation.
+ */
+// clang-format off
+#include <sparse/dgl_headers.h>
+// clang-format on
+
+#include <sparse/elementwise_op.h>
+#include <sparse/reduction.h>
+#include <sparse/sparse_matrix.h>
+#include <torch/script.h>
+
+#include <string>
+#include <vector>
+
+namespace dgl {
+namespace sparse {
+
+namespace {
+
+torch::Tensor ReduceAlong(
+    const c10::intrusive_ptr<SparseMatrix>& A, const std::string& reduce,
+    int64_t dim) {
+  auto value = A->value();
+  auto coo = A->COOPtr();
+
+  std::string reduce_op;
+  if (reduce == "sum") {
+    reduce_op = "sum";
+  } else if (reduce == "smin") {
+    reduce_op = "amin";
+  } else if (reduce == "smax") {
+    reduce_op = "amax";
+  } else if (reduce == "smean") {
+    reduce_op = "mean";
+  } else if (reduce == "prod") {
+    reduce_op = "prod";
+  } else {
+    TORCH_CHECK(false, "unknown reduce function ", reduce);
+    return torch::Tensor();
+  }
+
+  // Create the output tensor with shape
+  //
+  //   [A.num_rows if dim == 1 else A.num_cols] + A.val.shape[1:]
+  //
+  // (BarclayII) I'm not sure how to get the Tensor's shape as a vector...
+  auto value_shape = value.sizes();
+  std::vector<int64_t> output_shape(value_shape.begin(), value_shape.end());
+  std::vector<int64_t> view_dims(value_shape.size(), 1);
+  view_dims[0] = -1;
+  torch::Tensor idx;
+  if (dim == 0) {
+    output_shape[0] = coo->num_cols;
+    idx = coo->col.view(view_dims).expand_as(value);
+  } else if (dim == 1) {
+    output_shape[0] = coo->num_rows;
+    idx = coo->row.view(view_dims).expand_as(value);
```

Comment:
In the Python side, we do not need to expand `idx` to the shape of `value`. Maybe it is also not necessary in the C++ side? I am not very sure.